src/rust/ensogl/lib/components/src/file_browser.rs

```diff
@@ -0,0 +1,328 @@
+    // covers the toolbar of the file browser.
+    ensogl_core::define_shape_system! {
+        (style:Style) {
+            let sprite_width  : Var<Pixels> = "input_size.x".into();
+            let sprite_height : Var<Pixels> = "input_size.y".into();
+            let width         = sprite_width - SHADOW_PX.px() * 2.0;
+            let height        = sprite_height - SHADOW_PX.px() * 2.0;
+            let color         = style.get_color(theme::application::file_browser::background);
+            let rect          = Rect((&width,&height)).corners_radius(CORNER_RADIUS_PX.px());
+            let shape         = rect.fill(color);
+
+            let shadow  = shadow::from_shape(rect.into(),style);
+
+            let toolbar_border = Rect((width, TOOLBAR_BORDER_SIZE.px()));
+            let toolbar_border = toolbar_border.translate_y(height / 2.0 - TOOLBAR_HEIGHT.px());
+
+            let toolbar_border_color_theme = theme::application::file_browser::toolbar_border_color;
+            let toolbar_border_color       = style.get_color(toolbar_border_color_theme);
+            let toolbar_border             = toolbar_border.fill(toolbar_border_color);
+
+            (shadow + shape + toolbar_border).into()
+        }
+    }
+}
+
+
+
+// =============
+// === Model ===
+// =============
+
+#[derive(Clone,Debug)]
+struct Model {
+    app            : Application,
+    display_object : display::object::Instance,
+    columns        : RefCell<Vec<Column>>,
+    scroll_area    : ScrollArea,
+    background     : background::View,
+    title          : text::Area,
+    // Invariant: The column with this index (if it exists) has `focused` set, all others have it
+    //            unset. But the index does not have to refer to an existing column.
+    focused_column : Cell<usize>,
+}
+
+impl Model {
+    /// This requires that `index` points to an existing column.
+    fn focus_column(&self, index:usize) {
+        let old_index  = self.focused_column.replace(index);
+        let old_column = self.columns.borrow().get(old_index).cloned();
+        if let Some(old_column) = old_column {
+            old_column.defocus();
+        }
+        let new_column = self.columns.borrow().get(index).cloned();
+        let new_column = new_column.expect("Trying to focus non-existent column.");
+        new_column.focus();
+        self.scroll_to_column(&new_column);
+    }
+
+    fn move_focus_by(&self, amount:isize) {
+        let old_index  = self.focused_column.get();
+        let max_index  = self.columns.borrow().len() as isize - 1;
+        let new_index  = old_index as isize + amount;
+        let new_index  = new_index.max(0).min(max_index);
+        let new_index  = new_index as usize;
+        let new_column = self.columns.borrow()[new_index].clone();
+        let entries    = new_column.model.entries.borrow().as_ref().cloned();
+        if let Some(entries) = entries {
+            if entries.len() > 0 {
```

Comment:
I think I chose this because I instinctively avoid negation and "not empty" almost feels like a double negation. But I am okay with that way to formulate it.

Reply:
Hmm, well, actually there's no need, `> 0` means the same.